ci-services/circleci.js

```diff
@@ -1,13 +1,28 @@
 'use strict'
 
 const _ = require('lodash')
+const gitHelpers = require('../lib/git-helpers')
 
 const env = process.env
 
+/**
+ * Last commit is a lockfile update
+ */
+function isLockfileUpdate () {
+  const reUpdateLockfile = /^chore\(package\): update lockfile*$/mi
+  const lastCommitMessage = gitHelpers.getLastCommitMessage()
+  return reUpdateLockfile.test(lastCommitMessage)
+}
+
+// For use in workflows, the .git directory needs to be saved to cache after updating the lockfile
+// and the cache needs to be restored in a later job before uploading the lockfile
 module.exports = {
   repoSlug: `${env.CIRCLE_PROJECT_USERNAME}/${env.CIRCLE_PROJECT_REPONAME}`,
   branchName: env.CIRCLE_BRANCH,
-  firstPush: !env.CIRCLE_PREVIOUS_BUILD_NUM,
+  // CIRCLE_PREVIOUS_BUILD_NUM is null on the first job of the first workflow on the branch
+  // CIRCLE_WORKFLOW_UPSTREAM_JOB_IDS is null on the first job of every workflow on the branch
+  // (update on first job of first workflow) or (upload when not first job in workflow and last commit is lockfile update)
+  firstPush: !env.CIRCLE_PREVIOUS_BUILD_NUM || (env.CIRCLE_WORKFLOW_UPSTREAM_JOB_IDS && isLockfileUpdate()),
```

Comment:
I tried checking that there are two commits on the branch (the greenkeeper package update and the lockfile update), but this did not work. So, the `isLockfileUpdate()` regex check is necessary here.

Reply:
git-helpers getNumberOfCommitsOnBranch does not work on CircleCI and always returns 0.
https://github.com/patkub/test-gk-lock-git-helpers
I installed `greenkeeper-lockfile` as a submodule and ran:
```
const gitHelpers = require('./greenkeeper-lockfile/lib/git-helpers.js')
console.log('Number of commits on branch: ' + gitHelpers.getNumberOfCommitsOnBranch())
console.log('Last commit message:\n' + gitHelpers.getLastCommitMessage())
```
https://circleci.com/gh/patkub/test-gk-lock-git-helpers/tree/greenkeeper%2Ftest-greenkeeper-lockfile-appveyor-package-1.0.63